File: android/core/src/test/kotlin/com/stockbook/core/CreditNoteTests.kt
package com.stockbook.core

import com.stockbook.core.model.StatementPeriod
import com.stockbook.core.store.DraftLine
import com.stockbook.core.store.InMemoryRepository
import com.stockbook.core.store.StockbookStore
import com.stockbook.core.text.AppLanguage
import com.stockbook.core.text.Strings
import com.stockbook.core.transfer.BackupService
import java.time.Instant
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Money the shop gives back without money moving.
 *
 * The figures here are the whole feature. A credit note that reduced the balance
 * by the wrong amount, or that put stock back twice, is invisible on screen until
 * somebody counts a bin or asks a customer for money that was written off — so
 * every test below checks a **figure or a shelf count**, never merely that a call
 * returned non-null.
 */
class CreditNoteTests {

    private val english = Strings(AppLanguage.ENGLISH)

    private fun store() = StockbookStore(InMemoryRepository())

    private fun StockbookStore.aProduct() = addProduct("Cisa lock", 50, 60.0, 95.0)

    private fun StockbookStore.aCustomerOwing(): String {
        addCustomer("Ahmed")
        // 2000 billed, 0 paid — Ahmed owes 2000.
        saveBill(customer = "Ahmed", paid = 0.0, amount = 2000.0, invoiceNo = "1001")
        return "ahmed"
    }

    // --- What it does to the balance

    @Test
    fun `a credit note reduces what the customer owes`() {
        val store = store()
        val key = store.aCustomerOwing()
        assertEquals(2000.0, store.customer(key)?.owed)

        store.addCreditNote(customerKey = key, amount = 540.0, noteNo = "00130")

        assertEquals(1460.0, store.customer(key)?.owed, "2000 billed less 540 credited")
    }

    @Test
    fun `a credit note is not a payment`() {
        // The distinction the whole type exists for: both reduce the balance and
        // only one of them is cash. A statement that conflated them would tell
        // the owner they had taken money they never took.
        val store = store()
        val key = store.aCustomerOwing()

        store.addCreditNote(customerKey = key, amount = 540.0, noteNo = "00130")

        val statement = assertNotNull(store.statementForCustomer(key, StatementPeriod.thisMonth()))
        assertEquals(0.0, statement.received, "no money arrived")
        assertEquals(540.0, statement.credited)
        assertEquals(2000.0, statement.billed, "and the invoice still says what it said")
        assertEquals(1460.0, statement.closingBalance)
    }

    @Test
    fun `credit and cash both come off the closing balance`() {
        val store = store()
        val key = store.aCustomerOwing()

        store.recordPayment(key, 300.0)
        store.addCreditNote(customerKey = key, amount = 200.0, noteNo = "00130")

        val statement = assertNotNull(store.statementForCustomer(key, StatementPeriod.thisMonth()))
        assertEquals(300.0, statement.received)
        assertEquals(200.0, statement.credited)
        assertEquals(1500.0, statement.closingBalance, "2000 − 300 − 200")
        assertEquals(1500.0, store.customer(key)?.owed, "and the roster agrees with the document")
    }

    @Test
    fun `a credit note appears on the statement in date order`() {
        val store = store()
        store.addCustomer("Ahmed")
        store.saveBill(
            customer = "Ahmed",
            paid = 0.0,
            amount = 1000.0,
            createdAt = Instant.parse("2026-08-01T09:00:00Z"),
            invoiceNo = "1001"
        )
        store.addCreditNote(
            customerKey = "ahmed",
            amount = 250.0,
            noteNo = "00130",
            issuedAt = Instant.parse("2026-08-05T09:00:00Z")
        )

        val statement = assertNotNull(
            store.statementForCustomer("ahmed", StatementPeriod.Month(Instant.parse("2026-08-10T00:00:00Z")))
        )

        assertEquals(2, statement.entries.size)
        assertEquals(listOf(1000.0, 750.0), statement.runningBalances, "the balance column reads down")
    }

    @Test
    fun `a credit note from an earlier month is carried forward`() {
        // The reason a statement has an opening balance at all: a customer
        // credited in July must not look like they owe it again in August.
        val store = store()
        store.addCustomer("Ahmed")
        store.saveBill(
            customer = "Ahmed",
            paid = 0.0,
            amount = 1000.0,
            createdAt = Instant.parse("2026-07-10T09:00:00Z"),
            invoiceNo = "1001"
        )
        store.addCreditNote(
            customerKey = "ahmed",
            amount = 400.0,
            noteNo = "00130",
            issuedAt = Instant.parse("2026-07-20T09:00:00Z")
        )

        val august = assertNotNull(
            store.statementForCustomer("ahmed", StatementPeriod.Month(Instant.parse("2026-08-10T00:00:00Z")))
        )

        assertEquals(600.0, august.openingBalance)
        assertEquals(0.0, august.credited, "July's note belongs to July")
        assertEquals(600.0, august.closingBalance)
    }

    // --- The shelf

    @Test
    fun `returned items go back on the shelf`() {
        val store = store()
        val product = store.aProduct()
        store.addCustomer("Ahmed")
        store.saveBill(
            listOf(DraftLine(product.uid, 5, 95.0)),
            customer = "Ahmed",
            paid = 0.0,
            invoiceNo = "1001"
        )
        assertEquals(45, store.product(product.uid)?.stock, "the sale took five off")

        store.addCreditNote(
            customerKey = "ahmed",
            lines = listOf(DraftLine(product.uid, 2, 95.0)),
            noteNo = "00130"
        )

        assertEquals(47, store.product(product.uid)?.stock, "and two came back")
        assertEquals(190.0, store.creditNotes.first().total, "2 × 95, from the lines")
    }

    @Test
    fun `a credit note with no items moves no stock`() {
        // The mirror of a bill entered as a figure. An overcharge is not a return.
        val store = store()
        val product = store.aProduct()
        val key = store.aCustomerOwing()

        store.addCreditNote(customerKey = key, amount = 300.0, noteNo = "00130")

        assertEquals(50, store.product(product.uid)?.stock)
    }

    @Test
    fun `editing a credit note down leaves the right count on the shelf`() {
        // The bug this pins: without taking the old note's goods back first,
        // editing 5 to 3 would leave 8 on the shelf instead of 3.
        val store = store()
        val product = store.aProduct()
        store.addCustomer("Ahmed")
        store.saveBill(
            listOf(DraftLine(product.uid, 10, 95.0)),
            customer = "Ahmed",
            paid = 0.0,
            invoiceNo = "1001"
        )
        assertEquals(40, store.product(product.uid)?.stock)

        val note = assertNotNull(
            store.addCreditNote(
                customerKey = "ahmed",
                lines = listOf(DraftLine(product.uid, 5, 95.0)),
                noteNo = "00130"
            )
        )
        assertEquals(45, store.product(product.uid)?.stock)

        store.updateCreditNote(
            id = note.id,
            customerKey = "ahmed",
            lines = listOf(DraftLine(product.uid, 3, 95.0)),
            noteNo = "00130",
            issuedAt = note.issuedAt
        )

        assertEquals(43, store.product(product.uid)?.stock, "40 on the shelf plus the 3 returned")
        assertEquals(285.0, store.creditNotes.first().total)
    }

    @Test
    fun `removing a credit note takes its goods back off the shelf`() {
        val store = store()
        val product = store.aProduct()
        store.addCustomer("Ahmed")
        store.saveBill(
            listOf(DraftLine(product.uid, 5, 95.0)),
            customer = "Ahmed",
            paid = 0.0,
            invoiceNo = "1001"
        )
        val note = assertNotNull(
            store.addCreditNote(
                customerKey = "ahmed",
                lines = listOf(DraftLine(product.uid, 2, 95.0)),
                noteNo = "00130"
            )
        )
        assertEquals(47, store.product(product.uid)?.stock)

        store.deleteCreditNote(note.id)

        assertEquals(45, store.product(product.uid)?.stock)
        assertEquals(475.0, store.customer("ahmed")?.owed, "and the credit is gone from the balance")
    }

    // --- What it refuses

    @Test
    fun `a credit note for nothing is refused`() {
        val store = store()
        val key = store.aCustomerOwing()

        assertNull(store.addCreditNote(customerKey = key, amount = 0.0, noteNo = "00130"))
        assertNull(store.addCreditNote(customerKey = key, amount = null, noteNo = "00130"))
        assertNull(store.addCreditNote(customerKey = "", amount = 100.0, noteNo = "00130"))
        assertTrue(store.creditNotes.isEmpty())
    }

    @Test
    fun `a note number already used is found, whatever case it was typed in`() {
        val store = store()
        val key = store.aCustomerOwing()
        store.addCreditNote(customerKey = key, amount = 100.0, noteNo = "CN-0130")

        assertNotNull(store.creditNoteWithNo(" cn-0130 "))
        assertNull(store.creditNoteWithNo("CN-0131"))
        assertNull(store.creditNoteWithNo(""), "an empty box is not a clash with every blank note")
    }

    @Test
    fun `a credit note does not clash with a bill on the same number`() {
        // Its own series. "#00130" in a credit-note run has nothing to do with
        // invoice 00130, and refusing it would be the app inventing a rule the
        // shop's paper does not have.
        val store = store()
        val key = store.aCustomerOwing()

        store.addCreditNote(customerKey = key, amount = 100.0, noteNo = "1001")

        assertNotNull(store.creditNoteWithNo("1001"))
        assertEquals(1, store.creditNotes.size)
        assertNotNull(store.billWithInvoiceNo("1001"), "and the bill keeps its own number")
    }

    @Test
    fun `a note being edited does not clash with itself`() {
        val store = store()
        val key = store.aCustomerOwing()
        val note = assertNotNull(store.addCreditNote(customerKey = key, amount = 100.0, noteNo = "00130"))

        assertNull(store.creditNoteWithNo("00130", exceptId = note.id))
    }

    @Test
    fun `a note with no number is called what it is`() {
        val store = store()
        val key = store.aCustomerOwing()
        val note = assertNotNull(store.addCreditNote(customerKey = key, amount = 100.0))

        assertNull(note.noteNo)
        assertEquals(english.creditNoteLabel, note.reference(english))
    }

    // --- The file

    @Test
    fun `credit notes survive a backup round trip`() {
        val store = store()
        val product = store.aProduct()
        store.addCustomer("Ahmed")
        store.saveBill(
            listOf(DraftLine(product.uid, 5, 95.0)),
            customer = "Ahmed",
            paid = 0.0,
            invoiceNo = "1001"
        )
        store.addCreditNote(
            customerKey = "ahmed",
            lines = listOf(DraftLine(product.uid, 2, 95.0)),
            noteNo = "00130",
            reason = "returned, damaged"
        )

        val document = BackupService.decode(BackupService.encode(store.makeBackupDocument()))

        // A reader that dropped these would show every credited customer owing
        // more than they do, and send the owner to ask for money written off
        // weeks ago. That is a figure misread, which is what bumps the version.
        assertEquals(3, document.version)

        val restored = store()
        restored.replaceEverything(document)

        val note = restored.creditNotes.single()
        assertEquals("00130", note.noteNo)
        assertEquals(190.0, note.total)
        assertEquals("returned, damaged", note.reason)
        assertEquals(1, note.lines.size)
        assertEquals(285.0, restored.customer("ahmed")?.owed, "475 billed less 190 credited")
    }

    @Test
    fun `a file with no credit notes still reads`() {
        // Nothing in a fresh shop has one, and the key is absent rather than
        // empty in a file written before they existed.
        val document = BackupService.decode(
            """
            {
              "version": 3,
              "exportedAt": "2026-08-18T09:00:00Z",
              "ownerName": "Khalid",
              "currencyCode": "SAR"
            }
            """.trimIndent()
        )

        assertTrue(document.creditNotes.isEmpty())
    }
}

File: ios/StockbookTests/CreditNoteTests.swift
import Testing
import Foundation
@testable import Stockbook

/// Money the shop gives back without money moving.
///
/// The twin of `CreditNoteTests.kt`, test for test. The figures here are the
/// whole feature: a credit note that reduced the balance by the wrong amount, or
/// that put stock back twice, is invisible on screen until somebody counts a bin
/// or asks a customer for money that was written off — so every test below checks
/// a **figure or a shelf count**, never merely that a call returned non-nil.
@Suite("Credit notes")
@MainActor
struct CreditNoteTests {

    private let english = Strings(language: .english)

    private func makeStore() -> StockbookStore {
        StockbookStore(repository: InMemoryRepository())
    }

    /// Fixed dates rather than offsets from now, so the month boundaries these
    /// tests lean on cannot drift with the day the suite is run.
    private var calendar: Calendar {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(secondsFromGMT: 0)!
        return calendar
    }

    private func date(_ year: Int, _ month: Int, _ day: Int) -> Date {
        calendar.date(from: DateComponents(year: year, month: month, day: day, hour: 12))!
    }

    @discardableResult
    private func aProduct(in store: StockbookStore) -> Product {
        store.addProduct(name: "Cisa lock", stock: 50, cost: 60, price: 95)
    }

    /// 2000 billed, nothing paid — Ahmed owes 2000.
    private func aCustomerOwing(in store: StockbookStore) -> String {
        store.addCustomer(name: "Ahmed")
        store.saveBill(customer: "Ahmed", paid: 0, amount: 2000, invoiceNo: "1001")
        return "ahmed"
    }

    // MARK: What it does to the balance

    @Test("A credit note reduces what the customer owes")
    func reducesBalance() {
        let store = makeStore()
        let key = aCustomerOwing(in: store)
        #expect(store.customer(key: key)?.owed == 2000)

        store.addCreditNote(customerKey: key, amount: 540, noteNo: "00130")

        #expect(store.customer(key: key)?.owed == 1460, "2000 billed less 540 credited")
    }

    @Test("A credit note is not a payment")
    func notAPayment() throws {
        // The distinction the whole type exists for: both reduce the balance and
        // only one of them is cash. A statement that conflated them would tell
        // the owner they had taken money they never took.
        let store = makeStore()
        let key = aCustomerOwing(in: store)

        store.addCreditNote(customerKey: key, amount: 540, noteNo: "00130")

        let statement = try #require(store.statement(forCustomer: key, period: .thisMonth()))
        #expect(statement.received == 0, "no money arrived")
        #expect(statement.credited == 540)
        #expect(statement.billed == 2000, "and the invoice still says what it said")
        #expect(statement.closingBalance == 1460)
    }

    @Test("Credit and cash both come off the closing balance")
    func creditAndCash() throws {
        let store = makeStore()
        let key = aCustomerOwing(in: store)

        store.recordPayment(customerKey: key, amount: 300)
        store.addCreditNote(customerKey: key, amount: 200, noteNo: "00130")

        let statement = try #require(store.statement(forCustomer: key, period: .thisMonth()))
        #expect(statement.received == 300)
        #expect(statement.credited == 200)
        #expect(statement.closingBalance == 1500, "2000 − 300 − 200")
        #expect(store.customer(key: key)?.owed == 1500, "and the roster agrees with the document")
    }

    @Test("A credit note appears on the statement in date order")
    func inDateOrder() throws {
        let store = makeStore()
        store.addCustomer(name: "Ahmed")
        store.saveBill(
            customer: "Ahmed",
            paid: 0,
            amount: 1000,
            createdAt: date(2026, 8, 1),
            invoiceNo: "1001"
        )
        store.addCreditNote(
            customerKey: "ahmed",
            amount: 250,
            noteNo: "00130",
            issuedAt: date(2026, 8, 5)
        )

        let statement = try #require(
            store.statement(forCustomer: "ahmed", period: .month(date(2026, 8, 10)))
        )

        #expect(statement.entries.count == 2)
        #expect(statement.runningBalances == [1000, 750], "the balance column reads down")
    }

    // MARK: The shelf

    @Test("Returned items go back on the shelf")
    func returnedItemsComeBack() throws {
        let store = makeStore()
        let product = aProduct(in: store)
        store.addCustomer(name: "Ahmed")
        store.saveBill(
            lines: [DraftLine(productUID: product.uid, qty: 5, price: 95)],
            customer: "Ahmed",
            paid: 0,
            invoiceNo: "1001"
        )
        #expect(store.product(uid: product.uid)?.stock == 45, "the sale took five off")

        store.addCreditNote(
            customerKey: "ahmed",
            lines: [DraftLine(productUID: product.uid, qty: 2, price: 95)],
            noteNo: "00130"
        )

        #expect(store.product(uid: product.uid)?.stock == 47, "and two came back")
        #expect(store.creditNotes.first?.total == 190, "2 × 95, from the lines")
    }

    @Test("A credit note with no items moves no stock")
    func figureOnlyMovesNothing() {
        // The mirror of a bill entered as a figure. An overcharge is not a return.
        let store = makeStore()
        let product = aProduct(in: store)
        let key = aCustomerOwing(in: store)

        store.addCreditNote(customerKey: key, amount: 300, noteNo: "00130")

        #expect(store.product(uid: product.uid)?.stock == 50)
    }

    @Test("Editing a credit note down leaves the right count on the shelf")
    func editingDown() throws {
        // The bug this pins: without taking the old note's goods back first,
        // editing 5 to 3 would leave 8 on the shelf instead of 3.
        let store = makeStore()
        let product = aProduct(in: store)
        store.addCustomer(name: "Ahmed")
        store.saveBill(
            lines: [DraftLine(productUID: product.uid, qty: 10, price: 95)],
            customer: "Ahmed",
            paid: 0,
            invoiceNo: "1001"
        )
        #expect(store.product(uid: product.uid)?.stock == 40)

        let note = try #require(
            store.addCreditNote(
                customerKey: "ahmed",
                lines: [DraftLine(productUID: product.uid, qty: 5, price: 95)],
                noteNo: "00130"
            )
        )
        #expect(store.product(uid: product.uid)?.stock == 45)

        store.updateCreditNote(
            id: note.id,
            customerKey: "ahmed",
            lines: [DraftLine(productUID: product.uid, qty: 3, price: 95)],
            noteNo: "00130",
            issuedAt: note.issuedAt
        )

        #expect(store.product(uid: product.uid)?.stock == 43, "40 on the shelf plus the 3 returned")
        #expect(store.creditNotes.first?.total == 285)
    }

    @Test("Removing a credit note takes its goods back off the shelf")
    func removing() throws {
        let store = makeStore()
        let product = aProduct(in: store)
        store.addCustomer(name: "Ahmed")
        store.saveBill(
            lines: [DraftLine(productUID: product.uid, qty: 5, price: 95)],
            customer: "Ahmed",
            paid: 0,
            invoiceNo: "1001"
        )
        let note = try #require(
            store.addCreditNote(
                customerKey: "ahmed",
                lines: [DraftLine(productUID: product.uid, qty: 2, price: 95)],
                noteNo: "00130"
            )
        )
        #expect(store.product(uid: product.uid)?.stock == 47)

        store.deleteCreditNote(id: note.id)

        #expect(store.product(uid: product.uid)?.stock == 45)
        #expect(store.customer(key: "ahmed")?.owed == 475, "and the credit is gone from the balance")
    }

    // MARK: What it refuses

    @Test("A credit note for nothing is refused")
    func refusesNothing() {
        let store = makeStore()
        let key = aCustomerOwing(in: store)

        #expect(store.addCreditNote(customerKey: key, amount: 0, noteNo: "00130") == nil)
        #expect(store.addCreditNote(customerKey: key, amount: nil, noteNo: "00130") == nil)
        #expect(store.addCreditNote(customerKey: "", amount: 100, noteNo: "00130") == nil)
        #expect(store.creditNotes.isEmpty)
    }

    @Test("A note number already used is found, whatever case it was typed in")
    func clashFound() {
        let store = makeStore()
        let key = aCustomerOwing(in: store)
        store.addCreditNote(customerKey: key, amount: 100, noteNo: "CN-0130")

        #expect(store.creditNoteWithNo(" cn-0130 ") != nil)
        #expect(store.creditNoteWithNo("CN-0131") == nil)
        #expect(store.creditNoteWithNo("") == nil, "an empty box is not a clash with every blank note")
    }

    @Test("A credit note does not clash with a bill on the same number")
    func ownSeries() {
        // Its own series. "#00130" in a credit-note run has nothing to do with
        // invoice 00130, and refusing it would be the app inventing a rule the
        // shop's paper does not have.
        let store = makeStore()
        let key = aCustomerOwing(in: store)

        store.addCreditNote(customerKey: key, amount: 100, noteNo: "1001")

        #expect(store.creditNoteWithNo("1001") != nil)
        #expect(store.creditNotes.count == 1)
        #expect(store.billWithInvoiceNo("1001") != nil, "and the bill keeps its own number")
    }

    @Test("A note being edited does not clash with itself")
    func noSelfClash() throws {
        let store = makeStore()
        let key = aCustomerOwing(in: store)
        let note = try #require(store.addCreditNote(customerKey: key, amount: 100, noteNo: "00130"))

        #expect(store.creditNoteWithNo("00130", exceptId: note.id) == nil)
    }

    @Test("A note with no number is called what it is")
    func fallbackLabel() throws {
        let store = makeStore()
        let key = aCustomerOwing(in: store)
        let note = try #require(store.addCreditNote(customerKey: key, amount: 100))

        #expect(note.noteNo == nil)
        #expect(note.reference(english) == english.creditNoteLabel)
    }

    // MARK: The file

    @Test("Credit notes survive a backup round trip")
    func roundTrip() throws {
        let store = makeStore()
        let product = aProduct(in: store)
        store.addCustomer(name: "Ahmed")
        store.saveBill(
            lines: [DraftLine(productUID: product.uid, qty: 5, price: 95)],
            customer: "Ahmed",
            paid: 0,
            invoiceNo: "1001"
        )
        store.addCreditNote(
            customerKey: "ahmed",
            lines: [DraftLine(productUID: product.uid, qty: 2, price: 95)],
            noteNo: "00130",
            reason: "returned, damaged"
        )

        let document = try BackupService.decode(try BackupService.encode(store.makeBackupDocument()))

        // A reader that dropped these would show every credited customer owing
        // more than they do, and send the owner to ask for money written off
        // weeks ago. That is a figure misread, which is what bumps the version.
        //
        // Asserted as "at least the version that carried them" rather than as a
        // literal: what this test is about is that credit notes travel, and
        // pinning the number here means every later bump edits a credit-note
        // test to say something it was never about.
        #expect(document.version >= 3)
        #expect(document.version == BackupDocument.currentVersion)

        let restored = makeStore()
        restored.replaceEverything(with: document)

        let note = try #require(restored.creditNotes.first)
        #expect(restored.creditNotes.count == 1)
        #expect(note.noteNo == "00130")
        #expect(note.total == 190)
        #expect(note.reason == "returned, damaged")
        #expect(note.lines.count == 1)
        #expect(restored.customer(key: "ahmed")?.owed == 285, "475 billed less 190 credited")
    }
}
